doc/wg/core/notes/core-2020-06-12.md

```diff
@@ -0,0 +1,230 @@
+   point of sharing a slice is to allow two-way communication, we need to
+either trust the capsule to give us the buffer back or we need to have a
+mechanism in the kernel to ensure that?
+ * Johnathan: If you want userspace to be able to reuse the memory arbitrarily,
+   yes. But the kernel model is workable, userspace can work with it
+ * Amit: Right, so if you had to send multiple packets, you would need multiple
+   userspace buffers for it?
+ * Johnathan: Yeah, it causes lots of buffer allocation issues
+ * Amit: I think what you are pointing out which I had not considered before.
+   This is not an issue only from the perspective of breaking functionality,
+but even if you only ever reuse that buffer with the same capsule, userspace
+has to trust the capsule in order to trust *its own* type safety!
+ * Johnathan: My current plan is to change the libtock-rs API to make that not
+   a problem
+ * Sam: ...
+ * Amit: Basically right now you have to trust a capsule with more than just
+   the memory you have allowed to them, or there can be arbitrary UB in
+userspace
+ * Amit: This discussion was convincing to me that we should have a way to
+   enforce this in the kernel
+ * Amit: If we could enforce this in the kernel, and unallow was guaranteed to
+   succeed, would that fix these issues?
+ * Johnathan: Yes, and we would get a beautiful API
+ * Amit: What if it could fail, but was enforced if it did succeed? There might
+   be lots of userspace overhead?
+ * Johnathan: Right, then we would probably need two APIs
+ * Amit: Thinking through this, my sense is that if we do not reach all the way
+   down into the hardware (pass process buffers down to hardware) I think we
+should be able to write this such that unallow always succeeds.
+ * Sam: Does that cause bigger problems?
+ * Leon: I think that if we wanted to give hardware rights to the userspace
+   buffers with for instance DMA I think that it is still possible to make the
+call always succeed and if it doesn't succeed then that is the fault of the
+capsule and the capsule should block up until the DMA succeeds or is aborted.
+ * Amit: Oh that is a good point, there is no way for the app to fail if it
+   cant run
+ * Amit: I am more convinced after this discussion that this makes sense to
+   pursue. Any dissenting voices?
+ * Brad: As someone who cannot even compile libtock-rs, I am more interested in
+   the fact that this would make it harder to write poor capsules, because all
+of these things would be enforced behind a grant so you can't store ad-hoc
+process state in a capsule
+ * Amit: Sorry, what can you not do then?
+ * Brad: (explains again)
+ * Hudson: Brad is not dissenting, just saying he likes it for a different
+   reason
+ * Amit: Oh okay good, so we have consensus, so we should politely and
+   thankfully ask Guillaume to go ahead with a PR
+ * Leon: I have another approach, Can I open a PR with this other approach so
+   we can compare them both
```

Comment:
@lschuermann please go ahead if you have some draft PR to experiment with this, and/or some design sketch to share on #1905! On my side, I likely won't have time to work on a PR proposal this week.

Reply:
Thanks! I'll see what I can do, I don't know whether I'll make it this week either. Given that our proposals have only minor technical differences (from what I see the difference manifests itself rather in the usage semantics), if you have any idea for implementing i.e. the core kernel-held allow-table efficiently, I'd love to chat about that (e.g. on Slack).